streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedPeek.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.processor.StateStore;
+
+class MaterializedPeek<K, V, S extends StateStore> extends Materialized<K, V, S> {
```

Comment:
This is hideous, but it's the best I could think of to get a look at the name in Materialized (which is protected) without auto-setting it (which happens in `MaterializedInternal`'s constructor.

Reply:
Actually, I did have another thought, which was that `MaterializedInternal` could avoid setting the name in the constructor and instead provide a method to do it. We could pursue a strategy where all the names are assigned in the DSL classes, which would make it much easier to see when they are getting assigned.

But this seemed like a risky change, so I opted for the less invasive choice.

Reply:
Yeah, I agree with you that it would be a substantial refactoring.  IMHO better to start with a smaller change and we can re-visit the name generation down the road.

Reply:
A similar idea but not introducing a new class: we can add a new constructor of `MaterializedInternal` that only takes one parameter `Materialized` and do not try to create internal names at all, then after the object is created with this constructor, its `storeName()` may return null.

I think it is similar to what you have here but with fewer LOCs :P Besides, we should add comment on that constructor that it will be removed once we do not do this silly thing any more.

Reply:
I think if we go down that road, we may aw well complete the journey to my alternative proposal to add a method to MaterializedInternal to explicitly set the store name if there isn't one and remove that logic from the constructor. This can be implemented safely by adding the new constructor you proposed, and mechanically in lining the old one.

Basically, every call to `new MaterializedInternal...` would become:
```
mi = new MaterializedInternal(...);
mi.generateStoreNameIfNotSet();
```

Then we could judiciously remove the second line in cases like this one where we do not want it.

Reply:
@vvcephei sounds good.